qiskit/circuit/quantumcircuitdata.py

```diff
@@ -14,9 +14,112 @@
+    """A sequence of the qubits that the operation is applied to."""
+    clbits: Tuple[Clbit, ...]
+    """A sequence of the classical bits that this operation reads from or writes to."""
+
+    def __init__(
+        self,
+        operation: Instruction,
+        qubits: Iterable[Qubit] = (),
+        clbits: Iterable[Clbit] = (),
+    ):
+        self.operation = operation
+        self.qubits = tuple(qubits)
+        self.clbits = tuple(clbits)
+        self._legacy_format_cache = None
+
+    def copy(self) -> "CircuitInstruction":
+        """Return a shallow copy of the :class:`CircuitInstruction`."""
+        return self.__class__(
+            operation=self.operation,
+            qubits=self.qubits,
+            clbits=self.clbits,
+        )
+
+    def replace(
+        self,
+        operation: Optional[Instruction] = None,
+        qubits: Optional[Iterable[Qubit]] = None,
+        clbits: Optional[Iterable[Clbit]] = None,
+    ) -> "CircuitInstruction":
+        """Return a new :class:`CircuitInstruction` with the given fields replaced."""
+        return self.__class__(
+            operation=self.operation if operation is None else operation,
+            qubits=self.qubits if qubits is None else qubits,
+            clbits=self.clbits if clbits is None else clbits,
+        )
 
-from qiskit.circuit.exceptions import CircuitError
-from qiskit.circuit.instruction import Instruction
+    def __repr__(self):
+        return (
+            f"{type(self).__name__}("
+            f"operation={self.operation!r}"
+            f", qubits={self.qubits!r}"
+            f", clbits={self.clbits!r}"
+            ")"
+        )
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            # Ordered from fastest comparisons to slowest.
+            return (
+                self.clbits == other.clbits
+                and self.qubits == other.qubits
+                and self.operation == other.operation
+            )
+        if isinstance(other, tuple):
+            return self._legacy_format == other
+        return NotImplemented
+
+    # Legacy tuple-like interface support.
+    #
+    # For a best attempt at API compatibility during the transition to using this new class, we need
+    # the interface to behave exactly like the old 3-tuple `(inst, qargs, cargs)` if it's treated
+    # like that via unpacking or similar.  That means that the `parameters` field is completely
+    # absent, and the qubits and clbits must be converted to lists.
+
+    @property
+    def _legacy_format(self):
+        if self._legacy_format_cache is None:
+            # The qubits and clbits were generally stored as lists in the old format, and various
+            # places assume that they will certainly be lists.
+            self._legacy_format_cache = (self.operation, list(self.qubits), list(self.clbits))
+        return self._legacy_format_cache
+
+    def __getitem__(self, key):
+        return self._legacy_format[key]
```

Comment:
Did you ever test if it was faster to just do the if checking on the index in `__getitem__`. My assumption is it's slower averaged over multiple accesses but maybe faster on the fist one.

Reply:
I was optimising most heavily for the `__iter__` access, because by far the most common tuple-like access pattern of this object is `for instr, qargs, cargs in data`, which needs to access all the elements.  Since I wanted `__iter__` to be _precisely_ the same, I made the legacy tuple interface, and once that was in place, the simplest way to maintain perfect backwards compatibility with tuple was to pass this through - that also has the benefit of handling `slice` correctly.

There's also a further trick where `CircuitInstruction.qubits` is a tuple, but the legacy format has a list there.  I wanted to make sure you'd always get the same list instance each time you queried, and since we already needed a way to generate and cache the legacy tuple for `__iter__`, I passed this through.

I wasn't so desperately concerned with micro timings of just one single item, because I'm not sure if that access pattern ever occurs, and since we changed over the access patterns entirely within Terra, almost all uses go through the new forms anyway.